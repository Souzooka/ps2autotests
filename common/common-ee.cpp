#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
