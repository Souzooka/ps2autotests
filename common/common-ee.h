#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
