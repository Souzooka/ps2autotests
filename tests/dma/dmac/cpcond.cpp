#include <common-ee.h>
#include <assert.h>
#include <kernel.h>
#include <malloc.h>
#include <string.h>
#include "../dmaregs.h"
#include "../dmasend.h"
#include "../dmatags.h"

static volatile DMA::Channel *const toSPR = DMA::D9;

u32 getCpCond() {
	u32 result = 0;
	asm volatile (
		".set noreorder\n"
		"sync.l\n"
		"sync.p\n"
		"bc0t done\n"
		"addiu %0, $0, 1\n"
		"b done\n"
		"addiu %0, $0, 0\n"
		"done:\n"
		: : "r"(result)
	);
	return result;
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");
	
	u32 *buf = (u32 *)memalign(16, 16 * 1024);
	buf[0] = 0x01234567;
	buf[1] = 0x89ABCDEF;
	buf[2] = 0xDEADBEEF;
	buf[3] = 0x1337C0DE;
	
	u32 result = 0;
	
	result = getCpCond();
	printf("initial value: %d\n", result);
	
	*DMA::D_PCR = DMA::D_PCR_CPC9;
	
	result = getCpCond();
	printf("value after setting PCR.CPC: %d\n", result);
	
	DMA::SendSimple(toSPR, buf, 16 * 1024);
	
	result = getCpCond();
	printf("value after sending DMA: %d\n", result);
	
	*DMA::D_PCR &= ~DMA::D_PCR_CPC9;
	
	result = getCpCond();
	printf("value after clearing PCR.CPC: %d\n", result);
	
	*DMA::D_PCR = DMA::D_PCR_CPC9;

	result = getCpCond();
	printf("value after restoring PCR.CPC: %d\n", result);
	
	*DMA::D_STAT = DMA::D_STAT_CIS9;
	
	result = getCpCond();
	printf("value after resetting STAT.CIS: %d\n", result);
	
	free(buf);
	
	printf("-- TEST END\n");
	return 0;
}
