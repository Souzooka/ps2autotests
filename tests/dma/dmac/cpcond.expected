-- TEST BEGIN
initial value: 1
value after setting PCR.CPC: 0
value after sending DMA: 1
value after clearing PCR.CPC: 1
value after restoring PCR.CPC: 1
value after resetting STAT.CIS: 0
-- TEST END
